File: Medium/Minimize the Heights II/minimize-the-heights-ii.py
#User function Template for python3

class Solution:
    def getMinDiff(self, arr, n, k):
        # code here
        arr.sort()
        diff = arr[n-1]-arr[0]
        for i in range(1,n):
            if(arr[i]-k<0):
                continue
            mini = min(arr[0]+k,arr[i]-k)
            maxi = max(arr[i-1]+k,arr[n-1]-k)
            diff = min(diff,maxi-mini)
        result = diff
        return result


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    tc = int(input())
    while tc > 0:
        k = int(input())
        n = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.getMinDiff(arr, n, k)
        print(ans)
        tc -= 1

# } Driver Code Ends

File: Difficulty: Medium/Minimize the Heights II/minimize-the-heights-ii.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
  public:
    int getMinDiff(vector<int> &arr, int k) {
        // code here
        if(arr.empty()) return 0;
        sort(arr.begin(),arr.end());
        int n =arr.size();
        int md = arr[n-1]-arr[0];
        for(int i=1;i<n;i+=1) {
            if(arr[i]<k) continue;
            int mv=min(arr[0]+k,arr[i]-k);
            int mx = max(arr[n-1]-k,arr[i-1]+k);
            md=min(md,mx-mv);
        }
        return md;
        
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        int n, k;
        cin >> k;
        cin.ignore();
        vector<int> a, b, c, d;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int num;
        while (ss >> num)
            a.push_back(num);

        Solution ob;
        auto ans = ob.getMinDiff(a, k);
        cout << ans << "\n";
    }
    return 0;
}
// } Driver Code Ends